aws_signing_helper/tpm_signer.go

```diff
@@ -0,0 +1,345 @@
+package aws_signing_helper
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"crypto/x509"
+	"encoding/asn1"
+	"encoding/pem"
+	"errors"
+	"io"
+	"math/big"
+	"os"
+
+	tpm2 "github.com/google/go-tpm/tpm2"
+	tpmutil "github.com/google/go-tpm/tpmutil"
+)
+
+type tpm2_TPMPolicy struct {
+	CommandCode   int    `asn1:"explicit,tag:0"`
+	CommandPolicy []byte `asn1:"explicit,tag:1"`
+}
+
+type tpm2_TPMAuthPolicy struct {
+	Name   string           `asn1:"utf8,optional,explicit,tag:0"`
+	Policy []tpm2_TPMPolicy `asn1:explicit,tag:1"`
+}
+
+type tpm2_TPMKey struct {
+	Oid        asn1.ObjectIdentifier
+	EmptyAuth  bool                 `asn1:"optional,explicit,tag:0"`
+	Policy     []tpm2_TPMPolicy     `asn1:"optional,explicit,tag:1"`
+	Secret     []byte               `asn1:"optional,explicit,tag:2"`
+	AuthPolicy []tpm2_TPMAuthPolicy `asn1:"optional,explicit,tag:3"`
+	Parent     int
+	Pubkey     []byte
+	Privkey    []byte
+}
+
+var oidLoadableKey = asn1.ObjectIdentifier{2, 23, 133, 10, 1, 3}
+
+type TPMv2Signer struct {
+	cert      *x509.Certificate
+	certChain []*x509.Certificate
+	tpmData   tpm2_TPMKey
+	public    tpm2.Public
+	private   []byte
+}
+
+func handleIsPersistent(h int) bool {
+	return (h >> 24) == int(tpm2.HandleTypePersistent)
+}
+
+var primaryParams = tpm2.Public{
+	Type:       tpm2.AlgECC,
+	NameAlg:    tpm2.AlgSHA256,
+	Attributes: tpm2.FlagUserWithAuth | tpm2.FlagRestricted | tpm2.FlagDecrypt | tpm2.FlagFixedTPM | tpm2.FlagFixedParent | tpm2.FlagNoDA | tpm2.FlagSensitiveDataOrigin,
+	ECCParameters: &tpm2.ECCParams{
+		Symmetric: &tpm2.SymScheme{
+			Alg:     tpm2.AlgAES,
+			KeyBits: 128,
+			Mode:    tpm2.AlgCFB,
+		},
+		Sign: &tpm2.SigScheme{
+			Alg: tpm2.AlgNull,
+		},
+		CurveID: tpm2.CurveNISTP256,
+		KDF: &tpm2.KDFScheme{
+			Alg: tpm2.AlgNull,
+		},
+	},
+}
+
+// Returns the public key associated with this TPMv2Signer
+func (tpmv2Signer *TPMv2Signer) Public() crypto.PublicKey {
+	ret, _ := tpmv2Signer.public.Key()
+	return ret
+}
+
+// Closes this TPMv2Signer
+func (tpmv2Signer *TPMv2Signer) Close() {
+}
+
+func padPKCS1_15(input []byte, size uint) (output []byte, err error) {
+	var PKCS1_PAD_OVERHEAD = 11
+
+	if uint(len(input)+PKCS1_PAD_OVERHEAD) > size {
+		return nil, errors.New("Digest too large for RSA signature")
+	}
+
+	padlen := size - uint(len(input))
+	output = make([]byte, size)
+	output[0] = 0x00
+	output[1] = 0x01
+	var i uint
+	for i = 2; i < padlen-1; i++ {
+		output[i] = 0xff
+	}
+	output[padlen-1] = 0x00
+	copy(output[padlen:], input)
+	return output, nil
+}
+
+// Implements the crypto.Signer interface and signs the passed in digest
+func (tpmv2Signer *TPMv2Signer) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	var paths []string
+	tpmdev := os.Getenv("TPM_DEVICE")
+	if tpmdev != "" {
+		paths = append(paths, tpmdev)
```

Comment:
Minor, but should we include explicit sensible defaults (maybe only for Windows)? The documentation of `OpenTPM` states the following: 
```
// This function may also be invoked with no paths, as tpm2.OpenTPM(). In this
// case, the default paths on Linux (/dev/tpmrm0 then /dev/tpm0), will be used.
```

Reply:
I'm happier letting go-tpm get this right; didn't I only add the TPM_DEVICE support to make the testing using the sim work correctly?